File: projects/detail-page-ai/src/orchestration/reference-selector.mjs
const ROLE_BONUS = Object.freeze({
  product_full: 0.22,
  product_closeup: 0.20,
  cross_section: 0.20,
  packaging: 0.16,
  delivery: 0.12,
  farm: 0.10,
  proof_document: 0.18,
  review_capture: 0.14,
  design_reference: 0.05,
  unknown: 0,
});

function scoreReference(file) {
  const quality = Number(file.qualityScore ?? 0);
  const identity = Number(file.productIdentityScore ?? 0);
  const usefulness = Number(file.usefulnessScore ?? 0);
  const roleBonus = ROLE_BONUS[file.role] ?? 0;
  return quality * 0.35 + identity * 0.35 + usefulness * 0.30 + roleBonus;
}

/**
 * Selects a small diverse set of references. Similar images can share duplicateGroupId.
 */
export function selectReferences(files, maxCount) {
  const eligible = (files ?? [])
    .filter((file) => file && file.usable !== false)
    .map((file) => ({ ...file, _score: scoreReference(file) }))
    .sort((a, b) => b._score - a._score);

  const selected = [];
  const duplicateGroups = new Set();
  const selectedRoles = new Set();

  for (const file of eligible) {
    if (selected.length >= maxCount) break;
    if (file.duplicateGroupId && duplicateGroups.has(file.duplicateGroupId)) continue;
    if (selectedRoles.has(file.role)) continue;
    selected.push(file);
    if (file.duplicateGroupId) duplicateGroups.add(file.duplicateGroupId);
    selectedRoles.add(file.role);
  }

  for (const file of eligible) {
    if (selected.length >= maxCount) break;
    if (selected.some((picked) => picked.id === file.id)) continue;
    if (file.duplicateGroupId && duplicateGroups.has(file.duplicateGroupId)) continue;
    selected.push(file);
    if (file.duplicateGroupId) duplicateGroups.add(file.duplicateGroupId);
  }

  return selected.map(({ _score, ...file }) => file);
}

File: projects/detail-page-automation/scripts/reference_selector.py
from __future__ import annotations

from typing import Any

ROLE_BONUS = {
    "product_full": 0.22,
    "product_closeup": 0.20,
    "cross_section": 0.20,
    "packaging": 0.16,
    "delivery": 0.12,
    "farm": 0.10,
    "proof_document": 0.18,
    "review_capture": 0.14,
    "design_reference": 0.05,
    "unknown": 0.0,
}


def _score(file: dict[str, Any]) -> float:
    quality = float(file.get("quality_score", 0))
    identity = float(file.get("product_identity_score", 0))
    usefulness = float(file.get("usefulness_score", 0))
    return quality * 0.35 + identity * 0.35 + usefulness * 0.30 + ROLE_BONUS.get(str(file.get("role", "unknown")), 0)


def select_references(files: list[dict[str, Any]], max_count: int) -> list[dict[str, Any]]:
    """Prefer clear, product-faithful, role-diverse files and skip near-duplicate groups."""
    eligible = [dict(item) for item in files if item and item.get("usable", True) is not False]
    eligible.sort(key=_score, reverse=True)

    selected: list[dict[str, Any]] = []
    duplicate_groups: set[str] = set()
    roles: set[str] = set()

    for item in eligible:
        if len(selected) >= max_count:
            break
        group = item.get("duplicate_group_id")
        role = str(item.get("role", "unknown"))
        if group and group in duplicate_groups:
            continue
        if role in roles:
            continue
        selected.append(item)
        roles.add(role)
        if group:
            duplicate_groups.add(str(group))

    for item in eligible:
        if len(selected) >= max_count:
            break
        if any(picked.get("id") == item.get("id") for picked in selected):
            continue
        group = item.get("duplicate_group_id")
        if group and str(group) in duplicate_groups:
            continue
        selected.append(item)
        if group:
            duplicate_groups.add(str(group))

    return selected
